---
title: "Appendix PD vs Controls, Padlocks"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    includes:
      in_header: ../include/in_header.html
      before_body: ../include/before_body.html
      after_body: ../include/after_body.html
runtime: shiny_prerendered
---
<!-- # (c)  Juozas GordeviÄius -->

```{r, include=FALSE}
source("../code/common.R")
params <- list(
  host          = system("hostname", intern=TRUE),
  nPerm         = 40,
  workers       = 41,
  outliers      = list(
    thresholdSD = 3,
    noOfPCs     = 2,
    removeOutliers = TRUE,
    shape       = "Diagnosis"
  ),
  clustering    = list(
    groups      = c(Diagnosis="Diagnosis", Batch="Batch", Gender="Sex", Plate="substr(ID, 1, 1)"),
    addClustersToModel = 0
  ),
  normalization = list(
    normalize   = TRUE,
    plotColor   = c("Diagnosis", "Batch")
  ),
  imputation    = list(
    imputeBetas = FALSE,
    imputeCounts= FALSE
  ),
  model         = list(
    formula     = "~ Diagnosis + .",
    formula0    = "~ . - Diagnosis",
    svRange     = 1:6
  )
)
```

# Parameters

```{r, include=TRUE}
params %>% jsonlite::toJSON(pretty=TRUE, auto_unbox = TRUE)
```

# Data


```{r}
padlock <- function() {
  data <- list()
  # Data directory
  d <- "../input/PD_Appendix_padloc_01172019"

  # read the sample key
  data$sample <- fread(file.path(d, "All_factor_rmAging.txt"))
  setnames(data$sample, gsub(" ", "", colnames(data$sample)))
  data$sample[, Age := as.numeric(Age)]
  data$sample[, PMI := as.numeric(PMI)]
  data$sample[, Sex := as.factor(Sex)]
  data$sample[, Diagnosis := as.factor(Diagnosis)]
  data$sample[, Batch := as.factor(Batch)]
  setnames(data$sample, "Plate_ID", "ID")

  # read the beta values
  data$beta <- fread(file.path(d, "data_raw_target_methy_noWGA_aveReplicate_70p_rm0_rmCommonSNP_CG_rmAging_50P.txt"))  

  # Move probe annotation to genome
  data$genome <- data$beta[, 1:5, with=FALSE] %>% 
    .[, ID := paste(chr, SNP, sep="_")]
  setnames(data$genome, Hmisc::capitalize(colnames(data$genome)))
  data$beta <- data$beta[, -(1:5), with=FALSE] %>%
    as.matrix
  rownames(data$beta) <- data$genome$ID
  data$beta <- data$beta / 100

  # read the sample counts
  data$counts <- fread(file.path(d, "data_raw.txt"))

  # Select CGs only
  data$counts <- data$counts[type == "CG"]
  data$counts[, ID := paste(chr, SNP, sep="_")]

  # reduce counts to the Cs present in the beta file
  data$counts <- data$counts[ ID %in% data$genome$ID]

  # get separate matrices for U and M
  id <- data$counts$ID
  ume_idx <- grep("_UMe", colnames(data$counts))
  me_idx <- grep("_Me", colnames(data$counts))
  data$countsU <- data$counts[, ume_idx, with=FALSE] %>% as.matrix
  data$countsM <- data$counts[, me_idx, with=FALSE] %>% as.matrix
  data$counts  <- data$countsU + data$countsM


  # Fix all names
  rownames(data$counts) <- rownames(data$countsU) <- rownames(data$countsM) <- id
  colnames(data$counts) <- colnames(data$countsU) <- colnames(data$countsM) <- gsub("_UMe", "", colnames(data$counts))


  # average technical replicate counts
  avgReps <- function(counts, reps) {
    stopifnot(all(reps %in% colnames(counts)))
    counts[, reps[1] ] <- rowMeans(counts[, reps ]) %>% round
    counts[, !colnames(counts) %in% reps[-1]]
  }
  foo <- function(data, reps) {
    data$counts <- avgReps(data$counts, reps)
    data$countsU <- avgReps(data$countsU, reps)
    data$countsM <- avgReps(data$countsM, reps)
    return(data)
  }
  data <- foo(data, c("C5", "F9", "F12"))
  data <- foo(data, c("C6", "F10", "G1"))
  data <- foo(data, c("C7", "F11", "G2"))
  data <- foo(data, c("C11", "G3", "G7"))
  data <- foo(data, c("D7", "G4", "G8"))
  data <- foo(data, c("E7","G5", "G9"))
  data <- foo(data, c("F1", "G6"))
  data <- foo(data, c("X3649", "X3649.R1", "X3649.R2"))
  data <- foo(data, c("C10", "X3557"))

  # Now match samples
  data$sample[grepl("^3", ID), ID := paste0("X", ID)]
  colnames(data$beta)[ grepl("^3", colnames(data$beta))] <- paste0("X", colnames(data$beta)[ grepl("^3", colnames(data$beta))])
  stopifnot(all(colnames(data$beta) %in% data$sample$ID))
  stopifnot(all(colnames(data$beta) == data$sample$ID))

  o <- match(data$sample$ID, colnames(data$counts))
  data$counts <- data$counts[, o]
  data$countsU <- data$countsU[, o]
  data$countsM <- data$countsM[, o]
  stopifnot(all(colnames(data$counts) == data$sample$ID))  
  stopifnot(all(colnames(data$countsU) == data$sample$ID))  
  stopifnot(all(colnames(data$countsM) == data$sample$ID))

  # Now match loci
  stopifnot(all(data$genome$ID == rownames(data$beta)))  
  stopifnot(all(data$genome$ID == rownames(data$counts)))  

  # Model data
  data$modelData <- data$sample[, 
    list(
      Diagnosis = as.factor(Diagnosis) %>% relevel("Control"), 
      Age, Gender=Sex, PMI, 
        Cohort = Batch
    )]

  return(data)  
}

# touch()
data <- cache(foo=padlock, fname="data.RDS", verbose=TRUE)
```

Resulting data structure:

```{r, include=TRUE}
sapply(data, dim)
```



<!-- Run the processing pipeline -->
```{r, child="../code/padlockPipeline.Rmd"}
```

# Robust lmFit 


```{r, include=TRUE, echo=TRUE}
methylation <- data$beta
prefix <- "m6_"
modelparams <- list(
    allowStable = 0.5,
    useMvalues  = TRUE,
    fixInfinity = 0.001, # Fraction of change for 0% and 100% methylated locations
    downloadDataMatrix = TRUE
  )
```


```{r, child="../code/padlockLmFit_2.Rmd"}
```


# SessionInfo

```{r, include=TRUE}
sessionInfo()
```
