---
title: "Mice rAAV-a-syn, Padlocks"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    includes:
      in_header: ../include/in_header.html
      before_body: ../include/before_body.html
      after_body: ../include/after_body.html
runtime: shiny_prerendered
---
<!-- # (c)  Juozas GordeviÄius -->

```{r, include=FALSE}
source("../code/common.R")
params <- list(
  host          = system("hostname", intern=TRUE),
  nPerm         = 40,
  workers       = 41,
  outliers      = list(
    thresholdSD = 3,
    noOfPCs     = 2,
    removeOutliers = TRUE,
    shape       = "Group"
  ),
  clustering    = list(
    groups      = c(Treatment="Treatment", Plate="substr(ID, 1, 1)"),
    addClustersToModel = 0
  ),
  normalization = list(
    normalize   = FALSE,
    plotColor   = c("Treatment")
  ),
  imputation    = list(
    imputeBetas = FALSE,
    imputeCounts= FALSE
  ),
  model         = list(
    formula     = "~ Treatment + .",
    formula0    = "~ . - Treatment",
    contrasts   = NULL,
    svRange     = 1:4,
    nsv         = 0
  )
)
```

# Parameters

```{r, include=TRUE}
params %>% jsonlite::toJSON(pretty=TRUE, auto_unbox = TRUE)
```

# Data


```{r}
padlock <- function() {
  data <- list()
  # Data directory
  d <- "../input/Mice_Padlock_1262019"

  # read the sample key
  data$sample <- fread("sampleKey.csv")
  data$sample <- data$sample[Cohort == "cecal patch"]

  # read the beta values
  data$beta <- fread(file.path(d, "DSS_mice_target_methy_removeReplicate_70p_rm0_CG.txt"))  

  # Move probe annotation to genome
  data$genome <- data$beta[, 1:5, with=FALSE] %>% 
    .[, ID := paste(chr, SNP, sep="_")]
  data$beta <- data$beta[, -(1:5), with=FALSE] %>%
    as.matrix
  rownames(data$beta) <- data$genome$ID
  data$beta <- data$beta / 100

  # read the sample counts
  data$counts <- fread(file.path(d, "DSS_mice_target_methy_removeReplicate_70p_rm0_CG_raw.txt"))

  # Select CGs only
  data$counts <- data$counts[type == "CG"]
  data$counts[, ID := paste(chr, SNP, sep="_")]

  # reduce counts to the Cs present in the beta file
  data$counts <- data$counts[ ID %in% data$genome$ID]

  # get separate matrices for U and M
  id <- data$counts$ID
  ume_idx <- grep("\\.UMe", colnames(data$counts))
  me_idx <- grep("\\.Me", colnames(data$counts))
  data$countsU <- data$counts[, ume_idx, with=FALSE] %>% as.matrix
  data$countsM <- data$counts[, me_idx, with=FALSE] %>% as.matrix
  data$counts  <- data$countsU + data$countsM


  # Fix all names
  rownames(data$counts) <- rownames(data$countsU) <- rownames(data$countsM) <- id
  colnames(data$counts) <- colnames(data$countsU) <- colnames(data$countsM) <- gsub("\\.UMe", "", colnames(data$counts))

  # average technical replicate counts
  avgReps <- function(counts, reps) {
    stopifnot(all(reps %in% colnames(counts)))
    counts[, reps[1] ] <- rowMeans(counts[, reps ]) %>% round
    counts[, !colnames(counts) %in% reps[-1]]
  }
  foo <- function(data, reps) {
    data$counts <- avgReps(data$counts, reps)
    data$countsU <- avgReps(data$countsU, reps)
    data$countsM <- avgReps(data$countsM, reps)
    return(data)
  }
  data <- foo(data, c("D5", "E3", "E4"))
  data <- foo(data, c("A11", "E5", "E6"))
  data <- foo(data, c("B8", "E7", "E8"))
  data <- foo(data, c("B4", "E9"))

  # Now match samples
  data$beta <- data$beta[, data$sample$ID]
  data$counts <- data$counts[, data$sample$ID]
  data$countsU <- data$countsU[, data$sample$ID]
  data$countsM <- data$countsM[, data$sample$ID]
  stopifnot(all(colnames(data$beta) %in% data$sample$ID))
  stopifnot(all(colnames(data$beta) == data$sample$ID))
  stopifnot(all(colnames(data$counts) == data$sample$ID))  
  stopifnot(all(colnames(data$countsU) == data$sample$ID))  
  stopifnot(all(colnames(data$countsM) == data$sample$ID))

  # Now match loci
  stopifnot(all(data$genome$ID == rownames(data$beta)))  
  stopifnot(all(data$genome$ID == rownames(data$counts)))  

  # For consistency with older data
  data$modelData <- data$sample[, 
    list(Treatment = Treatment %>% as.factor %>% relevel("GFP"))]

  return(data)  
}

# touch()
data <- cache(foo=padlock, fname="data.RDS", verbose=TRUE)
```

Resulting data structure:

```{r, include=TRUE}
sapply(data, dim)
```


<!-- Run the processing pipeline -->
```{r, child="../code/padlockPipeline.Rmd"}
```

# Robust lmFit


50% stable (5 out 10), non extreme change, M values, robust limma eBayes

```{r, include=TRUE, echo=TRUE}
methylation <- data$beta
prefix <- "m6_"
modelparams <- list(
    allowStable = 0.5,
    useMvalues  = TRUE,
    fixInfinity = 0.001, # Fraction of change for 0% and 100% methylated locations
    downloadDataMatrix = TRUE
  )
```


```{r, child="../code/padlockLmFit_2.Rmd"}
```

# SessionInfo

```{r, include=TRUE}
sessionInfo()
```